File: backend/src/schema/template.ts
import { z } from "zod";

/**
 * Template JSON contract shared with the Android app.
 * All coordinates are in design units (the Figma frame's pixel space, origin top-left).
 * Colors are "#RRGGBB".
 */

const hex = z.string().regex(/^#[0-9A-F]{6}$/i, "color must be #RRGGBB");

export const StrokeSchema = z.object({
  color: hex,
  width: z.number().positive(),
  align: z.enum(["INSIDE", "CENTER", "OUTSIDE"]),
});

const Bounds = {
  x: z.number(),
  y: z.number(),
  w: z.number().nonnegative(),
  h: z.number().nonnegative(),
};

export const ImageElementSchema = z.object({
  type: z.literal("image"),
  id: z.string(),
  name: z.string(),
  ...Bounds,
  src: z.string(),
  opacity: z.number().min(0).max(1).default(1),
});

export const ShapeElementSchema = z.object({
  type: z.enum(["rect", "ellipse"]),
  id: z.string(),
  ...Bounds,
  fill: hex.nullable(),
  fillOpacity: z.number().min(0).max(1).default(1),
  stroke: StrokeSchema.nullable(),
  radius: z.number().nonnegative().default(0),
  /** The design's "selected day" marker; the app moves it to the weekday of the date. */
  weekdayHighlight: z.boolean().default(false),
});

export const LineElementSchema = z.object({
  type: z.literal("line"),
  id: z.string(),
  x1: z.number(),
  y1: z.number(),
  x2: z.number(),
  y2: z.number(),
  color: hex,
  width: z.number().positive(),
});

export const CheckboxElementSchema = z.object({
  type: z.literal("checkbox"),
  id: z.string(),
  ...Bounds,
  fill: hex.nullable(),
  stroke: StrokeSchema.nullable(),
  checked: z.boolean(),
  shape: z.enum(["ellipse", "rect"]).default("ellipse"),
  radius: z.number().nonnegative().default(0),
  /** Check mark image, positioned relative to the checkbox origin. */
  check: z.object({ src: z.string(), dx: z.number(), dy: z.number(), w: z.number(), h: z.number() }),
});

export const TextStyleSchema = z.object({
  fontFamily: z.string(),
  fontWeight: z.number().int().min(100).max(900),
  italic: z.boolean(),
  fontSize: z.number().positive(),
  lineHeight: z.number().positive(),
  letterSpacing: z.number(),
  color: hex,
  align: z.enum(["LEFT", "CENTER", "RIGHT"]),
});

export const TextRunSchema = z.object({
  start: z.number().int().nonnegative(),
  end: z.number().int().positive(),
  fontWeight: z.number().int().optional(),
  fontFamily: z.string().optional(),
  italic: z.boolean().optional(),
});

export const TextElementSchema = z.object({
  type: z.literal("text"),
  id: z.string(),
  name: z.string(),
  text: z.string(),
  /** Original Figma text bounds (before rotation is applied). */
  ...Bounds,
  /** Writable area: user text wraps / ellipsizes inside this box. */
  box: z.object(Bounds),
  maxLines: z.number().int().positive(),
  style: TextStyleSchema,
  runs: z.array(TextRunSchema).default([]),
  list: z.enum(["none", "ordered", "unordered"]).default("none"),
  /** Number of the first ordered-list item (Figma continues numbering across stacked rows). */
  listStart: z.number().int().positive().default(1),
  /** Clockwise rotation in degrees around the box center. */
  rotation: z.number().default(0),
  /** "time" opens a time picker, "date" a date picker; both keep the original format. */
  role: z.enum(["text", "time", "date"]).default("text"),
  /** java.time DateTimeFormatter pattern used for role time/date. */
  format: z.string().optional(),
  /** Literal prefix kept in front of a date/time value, e.g. "Date: ". */
  prefix: z.string().default(""),
  uppercase: z.boolean().default(false),
  editable: z.boolean().default(true),
  /** Empty writing area generated from a ruled line or an empty box. */
  slot: z.boolean().default(false),
  /** Weekday letter (1 = Monday … 7 = Sunday); highlighted from the page's date, not editable. */
  weekday: z.number().int().min(1).max(7).optional(),
});

export const ElementSchema = z.discriminatedUnion("type", [
  ImageElementSchema,
  ShapeElementSchema.extend({ type: z.literal("rect") }),
  ShapeElementSchema.extend({ type: z.literal("ellipse") }),
  LineElementSchema,
  CheckboxElementSchema,
  TextElementSchema,
]);

export const TemplateSchema = z.object({
  id: z.string(),
  name: z.string(),
  category: z.string(),
  version: z.number().int().positive(),
  source: z.object({ fileKey: z.string(), nodeId: z.string(), syncedAt: z.string() }),
  page: z.object({
    width: z.number().positive(),
    height: z.number().positive(),
    background: hex,
    /** Output paper. The design is scaled uniformly to fit and centered on the sheet. */
    paper: z.object({ name: z.literal("A4"), widthPt: z.literal(595), heightPt: z.literal(842) }),
  }),
  editorOptions: z.object({
    colors: z.array(z.object({ name: z.string(), hex })),
    fonts: z.array(z.string()),
  }),
  elements: z.array(ElementSchema),
});

export type Template = z.infer<typeof TemplateSchema>;
export type TemplateElement = z.infer<typeof ElementSchema>;
export type TextElement = z.infer<typeof TextElementSchema>;

/** Plans store only what the user changed on top of a template. */
export const TextOverrideSchema = z.object({
  text: z.string().max(2000).optional(),
  color: hex.optional(),
  fontFamily: z.string().max(64).optional(),
});

export const PlanInputSchema = z.object({
  templateId: z.string().min(1),
  title: z.string().min(1).max(120),
  values: z.record(z.string(), TextOverrideSchema).default({}),
  checks: z.record(z.string(), z.boolean()).default({}),
});

export const ReminderInputSchema = z.object({
  title: z.string().min(1).max(120),
  note: z.string().max(500).default(""),
  remindAt: z.string().datetime({ offset: true }),
  enabled: z.boolean().default(true),
  planId: z.string().nullable().default(null),
  /** Text element the reminder was set from (task on a plan). */
  elementId: z.string().nullable().default(null),
});

File: frontend/app/src/main/java/com/dailyplanner/app/data/model/Template.kt
package com.dailyplanner.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Mirrors backend/src/schema/template.ts. Coordinates are Figma design units.
 */
@Serializable
data class Template(
    val id: String,
    val name: String,
    val category: String,
    val version: Int,
    val page: Page,
    val editorOptions: EditorOptions,
    val elements: List<Element>,
)

@Serializable
data class Page(
    val width: Float,
    val height: Float,
    val background: String,
    val paper: Paper,
)

@Serializable
data class Paper(val name: String, val widthPt: Int, val heightPt: Int)

@Serializable
data class EditorOptions(val colors: List<NamedColor>, val fonts: List<String>)

@Serializable
data class NamedColor(val name: String, val hex: String)

@Serializable
data class Stroke(val color: String, val width: Float, val align: String)

@Serializable
data class Box(val x: Float, val y: Float, val w: Float, val h: Float)

@Serializable
sealed interface Element {
    val id: String

    @Serializable
    @SerialName("image")
    data class Image(
        override val id: String,
        val name: String,
        val x: Float, val y: Float, val w: Float, val h: Float,
        val src: String,
        val opacity: Float = 1f,
    ) : Element

    @Serializable
    @SerialName("rect")
    data class Rect(
        override val id: String,
        val x: Float, val y: Float, val w: Float, val h: Float,
        val fill: String? = null,
        val fillOpacity: Float = 1f,
        val stroke: Stroke? = null,
        val radius: Float = 0f,
        /** Design's "selected day" marker, moved to the weekday of the page's date. */
        val weekdayHighlight: Boolean = false,
    ) : Element

    @Serializable
    @SerialName("ellipse")
    data class Ellipse(
        override val id: String,
        val x: Float, val y: Float, val w: Float, val h: Float,
        val fill: String? = null,
        val fillOpacity: Float = 1f,
        val stroke: Stroke? = null,
        val radius: Float = 0f,
        /** Design's "selected day" marker, moved to the weekday of the page's date. */
        val weekdayHighlight: Boolean = false,
    ) : Element

    @Serializable
    @SerialName("line")
    data class Line(
        override val id: String,
        val x1: Float, val y1: Float, val x2: Float, val y2: Float,
        val color: String,
        val width: Float,
    ) : Element

    @Serializable
    @SerialName("checkbox")
    data class Checkbox(
        override val id: String,
        val x: Float, val y: Float, val w: Float, val h: Float,
        val fill: String? = null,
        val stroke: Stroke? = null,
        val checked: Boolean,
        val shape: String = "ellipse",
        val radius: Float = 0f,
        val check: CheckMark,
    ) : Element

    @Serializable
    @SerialName("text")
    data class Text(
        override val id: String,
        val name: String,
        val text: String,
        val x: Float, val y: Float, val w: Float, val h: Float,
        val box: Box,
        val maxLines: Int,
        val style: TextStyle,
        val runs: List<TextRun> = emptyList(),
        val list: String = "none",
        val listStart: Int = 1,
        val rotation: Float = 0f,
        val role: String = "text",
        val format: String? = null,
        val prefix: String = "",
        val uppercase: Boolean = false,
        val editable: Boolean = true,
        /** Empty writing area generated from a ruled line or box. */
        val slot: Boolean = false,
        /** Weekday letter, 1 = Monday … 7 = Sunday. */
        val weekday: Int? = null,
    ) : Element
}

@Serializable
data class CheckMark(val src: String, val dx: Float, val dy: Float, val w: Float, val h: Float)

@Serializable
data class TextStyle(
    val fontFamily: String,
    val fontWeight: Int,
    val italic: Boolean,
    val fontSize: Float,
    val lineHeight: Float,
    val letterSpacing: Float,
    val color: String,
    val align: String,
)

@Serializable
data class TextRun(
    val start: Int,
    val end: Int,
    val fontWeight: Int? = null,
    val fontFamily: String? = null,
    val italic: Boolean? = null,
)
